test/helpers/kubectl.go

```diff
@@ -3805,10 +3805,14 @@ func (kub *Kubectl) ciliumServicePreFlightCheck() error {
 // Function needs a directory path where the files are going to be written
 // commands are run on all pods matching selector
 func (kub *Kubectl) reportMapContext(ctx context.Context, path string, reportCmds map[string]string, ns, selector string) {
+	pods, err := kub.GetPodNamesContext(ctx, ns, selector)
+	if err != nil {
+		log.WithError(err).Errorf("cannot retrieve pod names with '%s': %s", selector, err)
+	}
 	for cmd, logfile := range reportCmds {
-		results, err := kub.ExecInPods(ctx, ns, selector, cmd, true, ExecOptions{SkipLog: true})
-		if err != nil {
-			log.WithError(err).Errorf("cannot retrieve command output '%s': %s", cmd, err)
+		results := make(map[string]*CmdRes)
+		for _, pod := range pods {
+			results[pod] = kub.ExecPodCmdBackground(ctx, ns, pod, cmd, ExecOptions{SkipLog: true})
```

Comment:
`ExecPodCmdBackground` will not wait for the command to be finished, so we will end up writing incomplete output to report files.

I think that proper way of fixing this is changing `ExecInPods` so that it's runs all the commands in parallel and waits for all execs to finish.

Reply:
> `ExecPodCmdBackground` will not wait for the command to be finished, so we will end up writing incomplete output to report files.

Would adding calls to `WaitUntilFinish()` for each invocation of `ExecPodCmdBackground` here solve the problem of having incomplete results?

> I think that proper way of fixing this is changing `ExecInPods` so that it's runs all the commands in parallel and waits for all execs to finish.

A bit confused as from what I understand, the problem with calling `ExecInPods` is in the `wait` period where it waits for the pod to become ready and repeatedly do so for every invocation, hence parallelizing the cmd execution inside `ExecInPods` might not solve the problem?

Reply:
@nebril thoughts?

Reply:
PTAL :)

Reply:
@Weil0ng thanks for being patient. I meant that we should change implementation of `ExecInPods` so that each `ExecContext` call would be called in it's own goroutine, whole loop would be guarded by a waitgroup (`results` map would probably also need a lock to synchronize writing result from single call).

I am basically talking about rewriting this loop https://github.com/cilium/cilium/blob/3abfde24ea1a33693ca5a62c1fe32d7efce48b71/test/helpers/kubectl.go#L3153 to run goroutines inside and wait for them, similarly to what we do in `reportMapHost`: https://github.com/cilium/cilium/blob/3abfde24ea1a33693ca5a62c1fe32d7efce48b71/test/helpers/kubectl.go#L3831

Does this make sense?

Reply:
@nebril Thanks for the detailed explanation! I guess what I was trying to say is that even if we optimize that for-loop to multiple goroutines, it won't solve the original problem because the `wait` block above is the culprit here I think unless I missed something?

https://github.com/cilium/cilium/blob/3abfde24ea1a33693ca5a62c1fe32d7efce48b71/test/helpers/kubectl.go#L3145

Reply:
@Weil0ng `WaitForPods` runs `kubectl wait` underneath and is there only to make sure that all pods matching selector are ready before running actual commands.

Additionally to parallelizing the loop, it might be good idea to pass `wait=false` in `reportMapContext` because all the pods are usually already running and if they are not ready we may want to get logs from them anyway.

Reply:
@nebril The issue we're trying to address here is that `kubectl wait` gets run multiple times for the same pods. When those pods are not running (typically because cilium-agent fataled), that leads to long waiting times (~40min to collect logs, see #11976).

Reply:
@pchaigno I see, thanks for explaining! in that case it would make sense to just switch `wait` argument to `ExecInPods` to false in all logging functions, no point in waiting for pods when we are just gathering logs.

Reply:
Don't we deploy log-gatherer pods (or similar) that we would need to wait for at least once?

Reply:
We are waiting for log-gatherer pods when we are creating kubectl object: https://github.com/cilium/cilium/blob/3abfde24ea1a33693ca5a62c1fe32d7efce48b71/test/helpers/kubectl.go#L356

Reply:
Great, makes sense! Thanks a lot @nebril!

Reply:
Given that this is the only place `ExecInPods` is referenced, we might just get rid of the `wait` block entirely?

Reply:
@Weil0ng Sounds good to me :+1: